// nightwatch-api/lib/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.createSession = createSession;
Object.defineProperty(exports, "startWebDriver", {
  enumerable: true,
  get: function () {
    return _client.startWebDriver;
  }
});
Object.defineProperty(exports, "stopWebDriver", {
  enumerable: true,
  get: function () {
    return _client.stopWebDriver;
  }
});
Object.defineProperty(exports, "closeSession", {
  enumerable: true,
  get: function () {
    return _client.closeSession;
  }
});
exports.Section = exports.client = void 0;

var _client = require("./client");

var _promisify = require("./promisify");

var _proxy = _interopRequireDefault(require("./proxy"));

var _section = _interopRequireDefault(require("nightwatch/lib/page-object/section"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

let nightwatchClient;

async function createSession(env) {
  nightwatchClient = await (0, _client.createSession)(env);
  (0, _promisify.promisifyApi)(nightwatchClient, _client.runQueue);
  (0, _promisify.promisifyExpect)(nightwatchClient, _client.runQueue);
  (0, _promisify.promisifyPageObjects)(nightwatchClient, _client.runQueue);
}

const client = (0, _proxy.default)(() => nightwatchClient);
exports.client = client;

class Section extends _section.default {
  constructor(definition, options) {
    super(definition, options);
    (0, _promisify.promisifySection)(this.api, _client.runQueue);
  }

}

exports.Section = Section;
